Feature/list

xunxee opened the pull request:
## :: Actions Status

![CD preview](<https://github.com/xunxee/shopifyStore/actions/workflows/cd(preview).yml/badge.svg>)

[Preview](https://shopify-store-xunxee.vercel.app)

xunxee committed:
Fix: TitleContainer 컴포넌트에 unique key를 주지 않아 발생한 에러를 해결
- Item에 key 프로퍼티를 설정하여 에러를 해결

xunxee committed:
Fix: FooterTopList 컴포넌트에 unique key를 주지 않아 발생한 에러를 해결
- Item에 key 프로퍼티를 설정하여 에러를 해결

xunxee committed:
Add: CategoryBar 컴포넌트의 기본적인 렌더링을 테스트 및 구현
- CategoryBar 컴포넌트의 반복적인 코드를 줄이기 위해 map 함수를 사용

xunxee committed:
Add: fixtures 디렉터리에 listCategoriesStyle.js를 추가
- CategoryBar 컴포넌트에서 반복 사용될 CSS 스타일을 추가

xunxee committed:
Refactor: CategoryBar 컴포넌트에서 fixtures 디렉터리의 listCategoriesStyle을 사용하여 스타일링
- CategoryBar 컴포넌트에서 반복 사용될 CSS 스타일을 fixtures와 연결

xunxee committed:
Add: fixtures 디렉터리, listCategoriesStyle.js에 container 프로퍼티를 추가
- CategoryBar 컴포넌트에서 반복 사용될 CSS 스타일을 추가

xunxee committed:
Refactor: CategoryBar 컴포넌트에 fixtures 디렉터리, listCategoriesStyle의 container를 사용하여 스타일링
- CategoryBar 컴포넌트에서 반복 사용될 CSS 스타일을 fixtures와 연결

ChoJinmok commented on src/List/CategoryBar.jsx:
```diff
@@ -1,13 +1,46 @@
+  ...container,
+});
+
+const Layout = styled.ul({
+  ...layout,
+});
```
추가로 스타일을 더 줄 필요가 없다면

```javascript
const Layout = styled.ul(layout);
```

이런식으로 작성하시는게 더 좋을 거 같습니당!

xunxee committed:
Add: PR시 자동으로 PR을 작성해주는 템플릿을 추가
- CD 통과 여부를 알려주는 내용
- preview 사이트 URL

xunxee committed:
Fix: LoginPage 컴포넌트, refreshToken의 여부에 따라 보이는 보이는 화면 구성의 오류를 수정
- refreshToken이 없을 때는 Logo가, refreshToken이 있을 때는 LogoutPage가 보이도록 수정

xunxee committed:
Fix: slice.js, setIsAccountModalOpen 함수의 로직을 수정
- 모달창을 닫을 때 isLogin을 true로 설정하여, 모달창이 열릴 때는 무조건 로그인 화면이 렌더링되도록 수정

xunxee committed:
Add: fixtures 디렉터리, footerMenuList.js에 footer에 사용될 메뉴 리스트를 추가
- FooterTopList 컴포넌트와 테스트 코드에서 사용할 Footer 메뉴 리스트를 작성

xunxee committed:
Refactor: FooterTopList 컴포넌트와 테스트 코드에 fixtures, FOOTER_MENU_LIST를 연결
- FooterTopList 컴포넌트와 테스트 코드에서 공통으로 사용될 Footer 리스트를 연결하여 가독성과 보수성을 개선

xunxee committed:
Add: slice.js, clearPasswordInvalidCheckMessage 함수의 액션을 테스트 및 구현
- clearPasswordInvalidCheckMessage 함수는 password의 invalidCheckMessage를 초기화해주는 액션을 수행함
- 이를 가지고 회원가입시 password의 유효성 검사를 통과하면 password의 invalidCheckMessage를 초기화함

xunxee committed:
Add: fixtures 디렉터리, validFields.js를 추가
- email과 password의 유효성 검사를 위한 정규식 표현과 invalidMessage를 관리하기 위한 validFields.js를 생성

xunxee committed:
Add: HeaderPage 컴포넌트, handleToggle 함수 로직에 clearLoginFields를 추가
- 모달창이 닫히면, 입력되어있던 loginFields를 초기화하는 로직을 추가

xunxee committed:
Add: slice.js에 fixtures/validFields.js를 연결하여 값을 사용
- email과 password의 유효성 검사를 위한 정규식 표현과 invalidMessage를 fixtures에서 사용

xunxee committed:
Refactor: LoginForm에서 disabled를 결정하는 checkDisabled 함수와 handleChange 함수의 로직을 리팩터링
- state가 변할 때마다 checkDisabled 함수가 실행되는데, 이때 email이나 password의 invalidCehckMessage가 없을 때에만 로직이 실행되어 유효성 검사를 하도록 코드를 리팩터링
- onClick 이벤트가 발생하면 실행되는 handleChange 함수는, 유효성에 맞는 password를 입력시 password의 invalidMessage를 onClick 이벤트 중에 삭제하도록 로직을 추가

xunxee committed:
Refactor: fixtures, listCategoriesStyle.js의 layout을 button 태그로 수정
- 기존의 li 태그에서, onClick 이벤트를 위해서 button 태그로 수정
- button 태그로 수정되면서 필요한 CSS 설정을 추가
- 바뀐 listCategoriesStyle.js의 내용을 RelevanceBar 컴포넌트에 반영

xunxee committed:
Add: List/slice.js의 changesCategories 함수의 액션을 테스트 및 구현
- store.js에 List/slice.js를 연결

xunxee committed:
Add: CategoryBar 컴포넌트의 Categories를 클릭했을 때의 액션을 테스트 및 구현
- CategoryBar 컴포넌트에서 onClick 이벤트의 target 정보를 ListContainer 컴포넌트의 handleClickCategories 함수의 인자로 전달하여 함수를 실행

xunxee committed:
Add: slice.js의 clearEmailInvalidCheckMessage 함수의 액션을 구현 및 테스트
- onClick 이벤트 발생시 email의 유효성을 검사하여 통과시 invalidCheckMessage를 초기화하는 함수를 구현

xunxee committed:
Refactor: slice.js의 checkSignUpValid 함수의 로직을 수정
- checkSignUpValid 함수를 fixtures/validFields.js를 사용하여 코드를 가독성을 리팩터링

xunxee committed:
Refactor: 회원가입시 이메일과 비밀번호의 유효성 검사의 액션을 테스트 및 구현
- 회원가입 단계에서 성과 이름이 존재하고, 이메일이나 비밀번호의 유효성이 통과되면 invalidCheckMessage를 초기화하도록 액션을 리팩터링
- LoginFormContainer 컴포넌트에서는 이메일의 invalidCheckMessage를 초기화하는 로직을 추가

xunxee committed:
Add: List 페이지에서 카테고리를 클릭하면 밑줄이 표시되는 액션을 테스트 및 구현
- categories의 New Arrivals를 클릭하면 handleClickCategories 함수가 실행되어 클릭된 해당 카테고리 정보를 store에 저장하고, Button 컴포넌트의 underLine 속성 값을 삼항연산자의 결과를 통해  props로 전달받아 css를 추가하도록 설정

xunxee committed:
Add: fixtures/listCategoriesCollection.js를 생성
- List 페이지에서 사용되는 categories들의 목록을 모아두는 fixtures를 생성하여 관리
- CategoryBar 컴포넌트에서 해당 fixtures를 연결하여 사용

xunxee committed:
Add: slice.js의 changesProducts 함수의 액션을 테스트 및 구현

xunxee committed:
Add: slice.js의 changesSort 함수의 액션을 테스트 및 구현

xunxee committed:
Refactor: HeaderPage.test.jsx의 불필요한 테스트를 제거

xunxee committed:
Add: List 페이지에서 Sort 카테고리를 클릭하면 밑줄이 표시되는 액션을 테스트 및 구현
- RelevanceBar 컴포넌트에서 sort 카테고리의 목록을 클릭하면 handleClickSort 함수가 실행되며 타겟의 name을 onClickSort 함수의 인자로 넘겨준다.
- 인자를 넘겨받은 ListContainer 컴포넌트의 handleClickSort 함수는 store에 이를 저장하고, 인자의 이름과 같은 sort의 카테고리에 밑줄을 표시한다.

xunxee committed:
Add: slice.js의 changesMaterial 함수의 액션을 테스트 및 구현

xunxee committed:
Add: List 페이지에서 Material 카테고리를 클릭하면 밑줄이 표시되는 액션을 테스트 및 구현
- RelevanceBar 컴포넌트에서 material 카테고리의 목록을 클릭하면 handleClickMaterial 함수가 실행되며 타겟의 name을 onClickMaterial 함수의 인자로 넘겨준다.
- 인자를 넘겨받은 ListContainer 컴포넌트의 handleClickMaterial 함수는 store에 이를 저장하고, 인자의 이름과 같은 material의 카테고리에 밑줄을 표시한다.

xunxee committed:
Add: slice.js의 clearCategories 함수의 액션을 테스트 및 구현

xunxee committed:
Add: HeaderPage.jsx, handleClickCategories 함수의 로직을 추가
- Header의 All을 클릭시 categories 상태 값을 초기화하여 밑줄을 없애는 액션을 추가
- Header의 All을 제외한 메뉴를 클릭시 categories의 상태값을 추가하는 액션을 추가

xunxee committed:
Add: ListContainer 컴포넌트의 handleClickCategories 함수의 액션을 추가
- All Categories의 메뉴를 클릭시, useNavigate를 사용하여 URL을 추가

xunxee committed:
Modify: slice.js, initialState의 기본 state 명을 수정
- categories → category로 수정
- products → product로 수정

xunxee committed:
Modify: slice.js, reducers의 함수명 이름을 수정
- changesCategories → changeCategory로 수정
- clearCategories → clearCategory로 수정
- changesProducts → changeProduct로 수정
- changesSort → changeSort로 수정
- changesMaterial → changeMaterial로 수정

xunxee committed:
Modify: HeaderPage.jsx, slice의 바뀐 함수명 이름을 반영

ChoJinmok commented on src/List/CategoryBar.jsx:
```diff
@@ -1,43 +1,74 @@
-            { id: 2, title: 'Featured' },
-          ].map((item) => (
-            <Item key={item.id}>{item.title}</Item>
+          {LIST_CATEGORIES.categories.map((item) => (
```
`item` 매개변수가 직접적으로 사용이 안되면 바로 구조분해할당하셔도 좋을 거 같습니다 ㅎㅎ
다른 비슷한 부분도 확인해주세요!~

xunxee replied:
아! 이 부분 전에도 말씀해주셨던 부분인데 ㅠㅠ
말씀 다시 해주시니 기억나는군용
감사합니다🥹

ChoJinmok commented on src/List/RelevanceBar.jsx:
```diff
@@ -1,45 +1,74 @@
+    { target: { name } },
+  ) => {
+    onClickSort(name);
+  }, [sort]);
```
이 부분도 어제 같이 봤던 함수처럼 고쳐야겠네요?! 다른 함수들도 확인해보세용!

ChoJinmok commented on src/List/slice.js:
```diff
@@ -0,0 +1,46 @@
+    material: '',
+  },
+  reducers: {
+    changesCategories(state, { payload: value }) {
+      return {
+        ...state,
+        categories: value,
+      };
+    },
+    changesProducts(state, { payload: value }) {
+      return {
+        ...state,
+        products: value,
+      };
+    },
+    changesSort(state, { payload: value }) {
+      return {
+        ...state,
+        sort: value,
+      };
+    },
+    changesMaterial(state, { payload: value }) {
+      return {
+        ...state,
+        material: value,
+      };
+    },
```
이 부분과 `ListContainer.jsx`을 보면 비슷한 코드가 반복되는 거 같아서 reducer하나로 합치는게 좋지 않을까 합니다 ㅎㅎ

xunxee replied:
한 번에 모든 categories 들을 관리하는 reducer를 만들지, 아니면 잘게 나누어서 하나씩 다 만들지 고민했었는데,
역시 한 번에 다 해줄 수 있는게 좋겠군요!

ChoJinmok replied:
언제 나누고 언제 합칠지 건땅만의 기준을 생각해보시면 좋을 거 같습니다 ㅎㅎ

ChoJinmok commented on src/LoginPage/LoginForm.jsx:
```diff
@@ -48,9 +48,29 @@ export default memo(({
+        && handleInvalidCheckMessage(name);
+    }
+
+    return null;
```
if 문 두개가 name만 다르고 똑같은 코드기 때문에 함수로 묶어 주시면 좋을 거 같고~ handleChange함수의 리턴 값을 활용하는게 아니라면 마지막의 return문도 불필요해보이네요!~ 한번 확인해보시면 좋을 거 같습니다!~

xunxee replied:
중복되는 코드를 함수를 사용해서 한 번 리팩터링 해보아야 겠습니당! 감사합니다🙏🏼

ChoJinmok commented on src/LoginPage/LoginForm.test.jsx:
```diff
@@ -118,47 +118,109 @@ describe('LoginForm', () => {
-      const { getByPlaceholderText } = renderLoginForm({
-        isLogin: false,
-      });
+    describe('handleChange', () => {
+      context('when the email passes validation', () => {
+        it('execute the handleInvalidCheckMessage function', () => {
```
'handleChange execute the handleInvalidCheckMessage function when the email passes validation' 이런식으로 직접적으로 적어주기보다 상황을 적어주는 게 더 좋아보입니다!

```javascript
describe('change events', () => {
  context('with valid email', () => {
    it('clear email invalid message' () => {
    })
  })
})
```

이런식으루요? 참고하세용!~ 다른 테스트들도 확인해주세요!~

xunxee replied:
어느 순간부터 뭔가 테스트를 적을 때, 제가 작성하는 부분이 이상하다고 생각이 들었는데
이렇게 상황을 작성했어야 하는군요 ㅜㅜ
말씀해주신 부분들 생각해서 다시 적어보니 훨씬 자연스럽네요! 감사합니당👍

ChoJinmok commented on src/LoginPage/LoginForm.test.jsx:
```diff
@@ -210,34 +274,22 @@ describe('LoginForm', () => {
 
     inputs.forEach((input) => {
-      describe(input.invalidCheckMessage, () => {
+      describe(input, () => {
```
정확하게 상황을 적어주시고 싶으시면 
```javascriprt
describe(`${input} invalid check message`)
```
이렇게 해주시면 되겠죠??

ChoJinmok commented on src/LoginPage/slice.js:
```diff
@@ -115,6 +115,20 @@ const { actions, reducer } = createSlice({
+
     clearPasswordInvalidCheckMessage(state) {
       const { loginFields } = state;
       return {
```
이 두 reducer도 하나로 합치시는 건 어떤가용?~

xunxee replied:
이것도 바로 합쳐서 적용도 해보아야 겠습니다!
감사합니다👍

ChoJinmok commented on src/LoginPage/slice.js:
```diff
@@ -241,22 +256,17 @@ export function checkSignUpValid({ name, value }) {
-      },
-    };
+    if (name === 'email') {
+      return VALID_FIELDS.email.regexps.test(value)
+        ? '' : VALID_FIELDS.email.invalidMessage;
+    }
+
+    if (name === 'password') {
+      return VALID_FIELDS.password.regexps.test(value)
+        ? '' : VALID_FIELDS.password.invalidMessage;
+    }
 
-    return validChecks[name] ? validChecks[name]() : '';
+    return '';
```
```javascript
const { regexps, invalidMessage } = VALID_FIELDS[name];
const isValid = regexps.test(value);

return isValid ? '' : invalidMessage;
```

참고하세요 ㅎㅎ

xunxee replied:
아 VALID_FIELDS를 이렇게 이용할 수 있군요🥹
반복되는 코드들을 항상 더 신경써서 보아야 겠습니다
감사합니다!

xunxee committed:
Add: slice.js, changeAllCategories 함수의 액션을 테스트 및 구현
- initialState의 category, product, sort, material 모두를 관리할 수 있는 changeAllCategories 함수를 구현

xunxee committed:
Refactor: categoryBar 컴포넌트에서 LIST_CATEGORIES 사용시 구조분해할당을 사용하여 코드를 리팩터링
- LIST_CATEGORIES 사용시 구조분해할당을 사용하여, 기존의 item.name에서 → name으로 바로 사용가능하도록 코드를 리팩터링

xunxee committed:
Refactor: App 컴포넌트의 Route 경로를 수정

xunxee committed:
Refactor: HeaderPage 컴포넌트에 slice.js의 changeAllCategories 함수를 적용
- 기존의 change 함수에서 → changeAllCategories 함수로 변경

xunxee committed:
Refactor: TitleContainer 컴포넌트에서 props로 받는 함수의 이름을 변경
- TitleContainer의 부모 컴포넌트인 HeaderPage 컴포넌트에서 props로 받는 onClickCategories 함수의 이름을 onClick으로 변경

xunxee committed:
Add: ListContainer 컴포넌트에서 category와 product를 클릭할 때의 액션을 테스트 및 구현
- slice.js, changeAllCategories 함수를 사용하여 액션을 구현

xunxee committed:
Refactor: LoginForm 컴포넌트의 handleChange 함수의 로직을, 중첩 함수를 사용하여 리팩터링
- handleChange 함수 안에, 중복되는 if문을 checkValid 함수를 만들어 리팩터링

xunxee committed:
Refactor: LoginForm.test.jsx의 describe 설명을 추가

xunxee committed:
Add: Login/slice.js, clearInvalidCheckMessage 함수의 로직을 테스트 및 구현
- 기존의 clearEmailInvalidCheckMessage 함수와 clearPasswordInvalidCheckMessage 함수를 합쳐서 clearInvalidCheckMessage 함수를 구현

xunxee committed:
Refactor: LoginFormContainer 컴포넌트에 clearInvalidCheckMessage 함수를 적용

xunxee committed:
Refactor: LoginForm 컴포넌트에 handleChange 함수의 반복되는 로직을 리팩터링
- VALID_FIELDS를 사용하여 반복되는 코드를 리팩터링

xunxee committed:
Refactor: LoginForm 컴포넌트에서 name의 값에 따라 checkInput의 값을 동적으로 할당하여 반복되는 코드를 리팩터링

xunxee committed:
Refactor: slice.js, checkSignUpValid 함수의 로직을 VALID_FIELDS를 사용하여 리팩터링

xunxee committed:
Refactor: RelevanceBar 컴포넌트의 LIST_CATEGORIES를 구조분해할당하여 리팩터링

xunxee committed:
Add: ListContainer 컴포넌트의 sort를 클릭시 query parameter가 생성되는 액션을 테스트 및 구현
- useLocation을 사용하여 categories 및 products 클릭시에도 기존의 query parameter가 붙도록 설정

xunxee committed:
Add: fixtures/listCategoriesCollection.js에 headerCategories를 생성하여 TitleContainer 컴포넌트에서 이를 적용
- TitleContainer 컴포넌트에서 배열로 있던 categories를 fixtures를 만들어 사용

xunxee committed:
Refactor: fixtures/listCategoriesCollection.js에 headerCategories에 name을 추가

xunxee committed:
Refactor: HeaderPage 컴포넌트에서 categories를 클릭시 각 category를 state에 저장하는 액션을 리팩터링
- 기존에는 All 카테고리를 클릭시 initialState의 category를 초기화했지만 이제는 all을 값으로 저장하도록 설정

xunxee committed:
Add: TitleContainer 컴포넌트에서 categories를 클릭시 color를 주는 액션을 테스트 및 구현

xunxee committed:
Add: List/slice.js의 changeLightOnCategory 함수의 액션을 테스트 및 구현

ChoJinmok commented on src/Header/HeaderPage.jsx:
```diff
@@ -46,7 +51,15 @@ export default function HeaderPage() {
+
+    return name && dispatch(changeAllCategories({
+      name, belong: 'category',
+    }));
```
- `name`을 url에서 숫자로 끊기어서 기준을 나누는 것보다는 다른 조건을 한번 고민해보셔 좋을 거 같아요!~
- 함수에서`return`이 잘못사용되고 있어서 `return`을 어떻게 사용해야하는지 잡고 가시면 좋을 거 같습니다!

ChoJinmok commented on src/Header/HeaderPage.test.jsx:
```diff
@@ -72,6 +76,27 @@ describe('HeaderPage', () => {
       expect(mockUsedNavigate).toBeCalledWith('/search');
+
+      expect(dispatch).toBeCalledWith({
+        payload: undefined,
```
`payload`가 사용되지 않는 경우엔 그냥 전달하지 않으셔도됩니당!ㅎㅎ

ChoJinmok commented on src/List/CategoryBar.jsx:
```diff
@@ -21,53 +21,57 @@ const Button = styled.button(({ underLine }) => ({
+              key={id}
+              name={name}
+              underLine={category === name
                 ? 'underLine' : 'none'}
```
`underline`과 `none`은 css에서 사용되는 값이니까 props로 전달해주기 보다 styled안에서 사용하는게 좋을 거 같아요!

ChoJinmok commented on src/List/CategoryBar.jsx:
```diff
@@ -21,53 +21,57 @@ const Button = styled.button(({ underLine }) => ({
+              {title}
             </Button>
           ))}
         </Layout>
```
위의 카테고리도 마찬가지고 'RelevanceBar'에도 거의 동일한 `Layout`이 있어서 하나의 컴포넌트로 따로 만드시는게 좋을 거 같아요!

xunxee replied:
호오... CategoryBar 컴포넌트를 하나 새롭게 만들었는데
이렇게도 줄여지는군요 ㅜㅜ
그런데 아직 이렇게 컴포넌트로 구조를 쪼개는 것이 익숙하지 않아서
시간이 굉장히 오래 걸렸네요 🥹
리뷰 감사합니당

ChoJinmok commented on src/List/ListContainer.jsx:
```diff
@@ -49,29 +52,51 @@ export default function ListContainer() {
     key: 'material',
   }));
```
여기선 state를 따로따로 받아오는 것 보다는

```javascript
const { category, product, sort, material } = useSelect(({ list }) => list);
```

이런식으로 한번에 받는 게 더 좋을 거 같아요!

xunxee replied:
그렇네요
useSelector 사용할 때 get 함수 사용하는 것을 버릇처럼 다 사용했군요 ㅠㅠ

ChoJinmok commented on src/List/ListContainer.jsx:
```diff
@@ -49,29 +52,51 @@ export default function ListContainer() {
+      return;
+    }
+
+    navigate(`/search/${name}`);
```
아래의 `handleClickProduct`도 그렇고 아마 `RelevanceBar`에 전달되는 함수에도 비슷한 조건문이 들어가지 않을까 싶어요! 함수마다 조건문을 만들어주는 것보다는 **그 당시 state를 모두 고려해서 url 주소를 바꿔주는 함수**를 하나 만드셔서 각 함수에 넣어주시는 게 좋을 거 같아요!

ChoJinmok commented on src/List/ListContainer.jsx:
```diff
@@ -27,16 +28,18 @@ const Container = styled.div({
 });
 
 export default function ListContainer() {
+  const navigate = useNavigate();
```
이제 발견했는데 ㅠㅠ `navigate`나 `params` 등등 `react-router-dom`관련한 것들은 보통 **Page** 컴포넌트에 관심사를 분리해줍니다!!

xunxee replied:
이제 Header는 다행히 나누었는데 List가 남았군요 ㅠㅠ
감사합니다🙏🏼

ChoJinmok commented on src/List/ListContainer.test.jsx:
```diff
@@ -1,60 +1,157 @@
+    return render((
+      <MemoryRouter>
+        <ListContainer />
+      </MemoryRouter>
```
관심사를 분리한다면 Container 컴포넌트에 `MemoryRouter`가 필요하지 않게 됩니다!

ChoJinmok commented on src/List/ListContainer.test.jsx:
```diff
@@ -1,60 +1,157 @@
+      given('product', () => 'sofas');
+
+      it('generates at the end of the categories path', () => {
+        const { queryByText } = renderListContainer();
```
`queryByText`는 `not`을 함께사용할 때 보통 사용하고 그외의 경우에는 `getBy`가 좋은 거 같습니다! ( [참고](https://stackoverflow.com/questions/58857019/react-testing-library-avoid-getby) )
다른 테스트들도 한번 확인해보세요!

xunxee replied:
단순히 에러를 발생시키지 않아서 query를 사용했었는데
명확하게 대상을 찾을 때는 getBy를 사용해주고(혹시 없다면 에러를 통해 상세한 피드백을 받고)
없을 것이라고 판단되는 것은 queryBy를 사용해서 null을 반환받는게 좋겠군용

ChoJinmok commented on src/List/slice.js:
```diff
@@ -3,31 +3,34 @@ import { createSlice } from '@reduxjs/toolkit';
   },
   reducers: {
-    changesCategories(state, { payload: value }) {
+    changeAllCategories(
```
'changeProductDataField'? 이건 진짜 단어가 생각이 안나네요 😭

xunxee replied:
아니면 changeCategoriesDataField 이렇게 가려고 하네요 🥹

ChoJinmok commented on src/LoginPage/LoginForm.jsx:
```diff
@@ -48,29 +48,17 @@ export default memo(({
-    ) {
-      const emailValid = VALID_FIELDS.email.regexps
-        .test(value);
+    function checkValid() {
```
```javascript
function checkValid() {
  const passValidation = Object.entries(fileds).reduce(
    (isValid, [fieldName, {fieldValue, invalidCheckMessage}]) => {
      if (!isValid) return false;

      if (fieldName === name) {
        const { regexps } = VALID_FIELDS[name];

        return regexps.test(value);
      }

      if (fieldName === 'firstName' || fieldName === 'lastName') {
        return !!fieldValue;
      }

      return !(!fieldValue || invalidCheckMessage);
    }, true);
  
  if (passValidation) handleInvalidCheckMessage(name);
}

if (name === 'email' || name === 'password') checkValid();
```

xunxee committed:
Add: HeaderPage 컴포넌트와 HeaderContainer 컴포넌트를 분리하는 액션을 테스트 및 구현
- 기존의 HeaderPage 컴포넌트의 react-router-dom과 react-redux를 분리

xunxee committed:
Modify: TitleContainer.jsx를 TitleBar.jsx로 이름을 수정

xunxee committed:
Modify: SearchBarContainer.jsx를 SearchBar.jsx로 이름을 수정

xunxee committed:
Modify: PurchaseContainer.jsx를 UserBar.jsx로 이름을 수정

xunxee committed:
Refactor: CategoryBar 컴포넌트의 반환문에 있던 underLine의 삼항 조건 연산자를 styled로 이동

xunxee committed:
Modify: Fixtures/listCategoriesCollection.js의 데이터를 수정
- listCategoriesCollection 객체의 data에서 사용하지 않는 id를 모두 제거

xunxee committed:
Modify: fixtures/listCategoriesStyle.js의 container 객체를 수정
- 각 category의 위쪽과 왼쪽의 margin을 설정

xunxee committed:
Add: 기존의 ListContainer 컴포넌트에서 사용되던 비슷한 컴포넌트들을 CategoryBar 컴포넌트로 대체
- 공통적으로 사용되는 컴포넌트들을 하나의 컴포넌트로 사용가능하도록 CategoryBar.jsx를 추가

xunxee committed:
Refactor: ListContainer 컴포넌트의 useSelector 함수를 호출시 get 함수를 사용하지 않고 구조분해할당을 통해 state를 가지고 오도록 리팩터링

xunxee committed:
Modify: test 코드에서 queryBy로 되어 있던 테스트들을 getBy로 수정
- 명확하게 대상을 찾을 때는 getBy를 사용해주고(혹시 없다면 에러를 통해 상세한 피드백을 받고) 없을 것이라고 판단되는 것은 queryBy를 사용해서 null을 반환받는 것이 좋기 때문에 getBy로 테스트 코드를 수정

xunxee committed:
Add: Header/TitleBar 컴포넌트에 focus 이벤트의 액션을 테스트 및 구현

xunxee committed:
Modify: List/slice.js의 changeAllCategories 함수의 식별자 이름을 changeCategoriesDataField로 변경

xunxee committed:
Modify: List/slice.js의 changeAllCategories 함수의 식별자 이름을 changeCategoriesDataField로 변경

xunxee committed:
Refactor: List 페이지의 ListPage 컴포넌트와 ListContainer 컴포넌트의 의존성을 리팩터링
- ListPage 컴포넌트는 react-router-dom에 의존
- ListContainer 컴포넌트는 react-redux에 의존

xunxee committed:
Refactor: LoginForm 컴포넌트, 함수의 로직을 리팩터링
- checkDisabled 함수에서 예외처리가 모두 끝난 후, main 로직이 실행되도록 리팩터링
- handleChange 함수에서 잘못 사용되고 있던 return문의 구조를 리팩터링
- handleSubmit 함수에서 preventDefault를 바로 사용하도록 리팩터링

xunxee committed:
Add: LoginPage/slice.js의 checkInvalidMessageClear 함수의 액션을 테스트 및 구현

xunxee committed:
Refactor: LoginForm 컴포넌트의 handleChange 함수에서 일어나는 로직을 미들웨어로 나누어 리팩터링
- LoginFormContainer 컴포넌트의 handleChange 함수에, LoginForm 컴포넌트 handleChange 함수에서 사용되던 함수를 미들웨어로 만들고 이를 LoginFormContainer에 연결
- 미들웨어를 만들면서 LoginForm 컴포넌트의 handleChange 함수의 로직이 간결해지고, 렌더링에 집중하는 컴포넌트로 리팩터링

ChoJinmok commented on src/Header/HeaderContainer.jsx:
```diff
@@ -0,0 +1,101 @@
+export default function HeaderContainer({ onClick }) {
+  const dispatch = useDispatch();
+
+  const handleClickCategories = useCallback((
+    url,
+    name,
+  ) => {
```
이 부분도 RORO 패턴을 사용하시면 더 좋을 거 같아요!!

xunxee replied:
그렇군요
RORO 이용하겠습니다!

ChoJinmok commented on src/Header/TitleBar.jsx:
```diff
@@ -32,19 +34,23 @@ const Item = styled.li({
-}) {
+export default function TitleBar({ onClick }) {
   function handleClick() {
     return (event) => {
       event.preventDefault();
-      onClick(event.currentTarget.pathname);
+      onClick(
+        event.currentTarget.pathname,
+        event.currentTarget.name,
+      );
     };
   }
```
`onClick` 함수 매개변수를 구조분해하고 이런식으로 작성하면 코드가 간결해질 거 같아요!

```javascript
({ currentTarget: { name, pathname } }) => {
  onClick({ name, pathname });
}
```

xunxee replied:
위에서 preventDefault를 받아서
아래에서 구조분해할당 변수를 만들어 진행해야겠군용!
감사합니다🙏🏼

ChoJinmok commented on fixtures/listCategoriesCollection.js:
```diff
@@ -1,42 +1,55 @@
 const listCategoriesCollection = {
+  headerCategories: [
+    {
+      title: 'All',
+      name: 'all',
+      path: '/search',
+    },
+    {
+      title: 'New Arrivals',
+      name: 'new',
+      path: '/search/new',
+    },
+    {
+      title: 'Featured',
+      name: 'featured',
+      path: '/search/featured',
+    },
+  ],
```
path는 결국 '/search/`name`'이 되니까 따로 설정해주지 않아도 될 것 같습니다!
a태그에 href에 담아 줄 때 삼항조건연산자와 템플릿 문자열을 이용하시면 좋을 거 같네요 ㅎㅎ

xunxee replied:
그렇군요
이미 가지고 있는 재료로 충분히 가능한 부분이였네요
감사합니다🙏🏼

ChoJinmok commented on src/Header/HeaderContainer.jsx:
```diff
@@ -0,0 +1,101 @@
+  const isAccountModalOpen = useSelector(get({
+    page: 'login',
+    key: 'isAccountModalOpen',
+  }));
```
이 부분도 전에 리뷰 드렸던 것처럼 useSelect를 두번 사용할 필요 없이 한번에 두 state를 가지고올 수 있을 거 같아요!

ChoJinmok commented on src/List/ListContainer.test.jsx:
```diff
@@ -38,10 +32,16 @@ describe('ListContainer', () => {
+    { pathname, search } = {},
+  ) {
     return render((
       <MemoryRouter>
```
`useNavigate`를 사용하지 않게 됐으니 이제 `MemoryRouter`도 사용하지 않아도 됩니다!

ChoJinmok commented on src/List/component/CategoryBar.jsx:
```diff
@@ -0,0 +1,58 @@
+    <Container>
+      <div>
+        <Layout>
+          {title}
```
title은 ul 태그의 텍스트 컨텐트로 넣기 보다는 ul 태그 밖에서 span으로 감싸주는 것이 더 좋을 거 같아요!

ChoJinmok commented on src/List/ListContainer.jsx:
```diff
@@ -27,84 +20,107 @@ const Container = styled.div({
+          keyword="categories"
+          item={category}
+          onClick={handleClickCategory}
+        />
```
`title`, `keyword`가 연관성이 없이 원하는대로 값을 넣는 경우엔 지금처럼 나누는 것이 좋겠지만 현재의 경우에는  `title`, `keyword`가 연관성이 많은 것 같아서 저라면 다음과 같이 진행할 것 같습니다!

```javascript
// fixtures/listCategoriesCollection.js

categories: {
  title: 'All Categories',
  belong: 'category' // -> changeCategoriesDataField 액션에서 belong도 똑같은 로직으로 작성이 가능해집니다.
  data: [
      { subHeading: 'New Arrivals', value: 'new' },
      { subHeading: 'Featured', value: 'featured' },
  ],
}
```

```javascript
// ListContainer.jsx

const handleClick = useCallback(({ name, belong }) => {
  dispatch(changeCategoriesDataField({ name, belong }));
}, [dispatch]);

// ...

<CategoryBar
  field="categories"
  selectedItem={category}
  onClick={handleClick}
/>
```

```javascript
// CategoryBar.jsx

export default function CategoryBar({
  field,
  selectedItem,
  onClick,
}) {
  const { title, belong, data } = LIST_CATEGORIES[field];

  const handleClick = useCallback((
    { target: { name } },
  ) => {
    onClick({ name, belong });
  }, [item, onClick]);

  return (
    <Container>
      <div>
        <h1>{title}</h1>
        <Layout>
          {data.map(({ value, subHeading }) => (
            <Button
              type="button"
              key={value}
              name={value}
              underLine={item}
              onClick={handleClick}
            >
              {subHeading}
            </Button>
          ))}
        </Layout>
      </div>
    </Container>
  );
}
```

ChoJinmok commented on src/List/ListContainer.jsx:
```diff
@@ -27,84 +20,107 @@ const Container = styled.div({
+      name,
+      belong: 'material',
+    }));
   }, [dispatch]);
```
현재 상황에서는 `useLacation`의 `pathname`과 `search`가 불필요해보입니다!

지난번 리뷰에서 추가해서 도전해보셨으면 하는게 있어요!

제한 사항
1. `useLacation`의 `pathname`과 `search` 사용하지 않고 state만을 이용해서 url 만들기
2. 현재 주제별로 함수가 나누어져있는데 하나의 함수로 만들기
3. (가능하다면) 데이터를 다루는 기준들이 더 추가되어도 동작되게 하기

<br />

제가 직접 코드를 작성해보지 않아서 위의 조건이 절대 안될 수도 있어서 잘 안되시면 꼭 너무 고민해보지 마시고 말씀해주세요!

ChoJinmok commented on src/LoginPage/LoginForm.jsx:
```diff
@@ -21,54 +21,48 @@ export default memo(({
+    const passwordValidCheck = VALID_FIELDS.password.regexps
+      .test(password.value);
+
+    return (lastName.value && firstName.value
```
이 부분을 생각해보니까 28번째 줄에서 `lastName`과 `firstName`의 value 유무를 체크했기때문에 생략해도 될 거 같습니다!

ChoJinmok commented on src/LoginPage/LoginForm.jsx:
```diff
@@ -21,54 +21,48 @@ export default memo(({
 
-    return false;
+    const passwordValidCheck = VALID_FIELDS.password.regexps
+      .test(password.value);
```
이런식으로도 작성할 수 있지 않을까하네용?? 이건 참고만 하세용!

```javascript
const {
  email: { regexps: emailRegexps },
  password: { regexps: passwordRegexps }
} = VALID_FIELDS;

const emailValidCheck = emailRegexps.test(email.value);
const passwordValidCheck = passwordRegexps.test(password.value);

return emailValidCheck && passwordValidCheck;
```

ChoJinmok commented on src/LoginPage/slice.js:
```diff
@@ -264,4 +264,26 @@ export function checkSignUpValid({ name, value }) {
+      ? password : email;
+
+    return VALID_FIELDS[name].regexps.test(value)
+      && !checkInput.invalidCheckMessage;
```
이 조건이 해결하지 못하는 상황이 있을 거 같아요! 다음 상황일 때를 테스트 해봐야할 거 같아요!

1. firstName, lastName, email에 유효성 검사 메세지가 생긴 상황에서 email을 다시 제대로 입력하는 상황
2. email만 유효성 검사 메세지가 생긴 상황에서 email을 다시 제대로 입력하는 상황

xunxee committed:
Refactor: HeaderContainer 컴포넌트의 handleClickCategories 함수의 인자를 RORO 패턴으로 받도록 리팩터링
- TitleBar.test.jsx에 LIST_CATEGORIES를 사용하여 title을 테스트하도록 리팩터링

xunxee committed:
Refactor: TitleBar 컴포넌트, handleClick 함수의 인자로 받는 event 객체를 구조분해할당하여 사용

xunxee committed:
Refactor: TitleBar 컴포넌트에서 a태그의 href 속성을 줄 때 삼항조건연산자를 사용하여 값을 주도록 리팩터링
- fixtures/listCategoriesCollection의 path를 제거하고 name을 이용하여 TitleBar 컴포넌트에서 href의 값으로 사용하도록 리팩터링

xunxee committed:
Refactor: HeaderContainer 컴포넌트, useSelector에서 구조분해할당을 이용하여 리팩터링

xunxee committed:
Refactor: ListContainer 컴포넌트에서 react-router-dom을 사용하지 않으면서, 테스트 코드의 MemoryRouter를 제거

xunxee committed:
Refactor: fixtures/listCategoriesCollection.js의 데이터 형식을 객체로 리팩터링
- 기존 자료구조에서 객체를 추가하여 fixtures의 확장성을 리팩터링

xunxee committed:
Refactor: LoginPage/slice.js의 checkInvalidMessageClear 함수의 액션을 테스트 및 구현
- LoginFormContainer 컴포넌트의 handleChange 함수에서 checkInvalidMessageClear 함수를 실행
- LoginFormContainer 컴포넌트의 handleCheckSignUpValid 함수에서 checkSignUpValid 함수를 실행

xunxee committed:
Refactor: 로그인 모달창이 꺼질 때 버튼 활성화를 막아주기 위한 액션을 추가
- HeaderContainer 컴포넌트의 handleToggle 함수에 로직을 추가

xunxee committed:
Add: initialState, loginFields 안에 위치한 fieldState의 기본 형식을 담은  fixtures/fieldState.js를 추가
- slice의 액션을 테스트할 때 해당 fixtures를 활용

xunxee committed:
Refactor: checkInvalidMessageClear 함수에 반복문을 추가하는 액션을 테스트 및 구현
- validateRestSignUpFields 함수와 validateAccountFields 함수 로직을 추가

xunxee committed:
Add: 로그인 및 회원가입 모달창을 닫을 때 버튼 활성화를 다시 막아주는 액션을 추가

xunxee committed:
Add: slice.js, checkUrl 함수에서 사용할 url 객체를 추가

xunxee committed:
Add: slice.js, checkUrl 함수의 액션을 테스트 및 구현
- 사용자가 클릭한 카테고리를 state에 저장하고
- checkUrl 함수의 중첩 함수인 makePathname 함수와 makeSearch 함수를 사용하여 url 만든다.

xunxee committed:
Add: slice.js의 checkUrl 함수를 연결
- 이벤트가 발생하는 CategoryBar 컴포넌트에서 name과 belong을 ListContainer 컴포넌트의 함수로 전달
- ListContainer 컴포넌트는 CategoryBar 컴포넌트에서 받은 인자를 checkUrl로 전달하여 함수를 호출

xunxee committed:
Add: proxy 서버 설정을 위한 webpack devServer 설정과 vercel의 rewrites 설정을 추가

xunxee committed:
Refactor: Footer에서 container가 들어가 있는 컴포넌트의 이름을 리팩터링
- container 컴포넌트에서는 react-redux를 사용하지만 해당 Footer에서는 사용하지 않으므로 이를 리팩터링

xunxee committed:
Refactor: 로그인, 로그아웃, 회원가입을 포함할 수 있는 단어로 디렉터리와 파일명을 리팩터링
- Login 디렉터리명을 ControlMember로 리팩터링
- LoginPage 컴포넌트를 ControlMemberPage로 리팩터링

xunxee committed:
Refactor: LoginForm 컴포넌트의 이름을 ControlMemberForm으로 리팩터링

xunxee committed:
Refactor: ControlMember/slice.js의 initialState 변수명을 리팩터링
- slice의 name을 login에서 controlMember로 리팩터링
- loginFields를 accountFields로 리팩터링

xunxee committed:
Refactor: ControlMember/slice.js의 reducer 함수의 이름을 리팩터링
- changeLoginFields 함수명을 changeAccountFields로 변경
- changeLoginErrorMessage 함수명을 changeAccountErrorMessage로 변경
- clearLoginFields 함수명을 changeAccountFields로 변경

xunxee committed:
Refactor: List/slice.js의 changeUrlDataField 함수의 액션을 테스트 및 구현
- 기존의 initialState의 구조에서 url 객체를 추가
- initialState 구조의 변경으로 인해, changeUrlDataField 함수의 로직을 리팩터링

xunxee committed:
Refactor: List/ListContainer.jsx의 checkUrl 함수의 액션을 테스트 및 구현

xunxee committed:
Refactor: ControlMember/slice.js의 initialState에서 loginFields를 accountFields로 리팩터링

xunxee committed:
Refactor: ListContainer 컴포넌트의 checkUrl 함수의 로직을 리팩터링

ChoJinmok commented on src/ControlMember/ControlMemberContainer.test.jsx:
```diff
@@ -2,22 +2,22 @@ import { render, fireEvent } from '@testing-library/react';
+describe('ControlMemberContainer', () => {
   const dispatch = jest.fn();
 
+  useDispatch.mockImplementation(() => dispatch);
```
beforeEach 밖으로 빼주신 이유가 있을까요?~

xunxee replied:
mockImplementation도 안으로 넣어주었군요?

ChoJinmok replied:
- `useSelector`의 경우 테스트 마다 state가 달리해야하는 경우가 대부분이라 `beforeEach`와 `given`을 사용해서 각 테스트 마다 다른 state를 줄 수 있습니다!
- `useDispatch`의 경우 구현 사항(implementation)이 변화가 없어서 `beforeEach` 밖으로 빼줘도 테스트가 실행되는 데에는 이상이 없을겁니다~!(`dipatch.mockClear()`만 잘해주면 될겁니다!)

<br />

- 제가 느낀 `beforeEach`안에서 작성하는 이유는 `mockImplementation`들을 한군데 모음으로써 가독성이 높아지는 것 같아요!
- 그런데 `beforeEach`라는 건 결국 모든 테스트 전에 실행되는 코드인데 불필요하게 매번 `useDispatch`를 `mockImplementation`하기 때문에 이런것이 불필요하게 느껴지신다면 밖으로 빼주셔도 상관 없습니다! (아샬님도 `beforeEach`안으로 몰아주는 것으로 봐선 크게 자원이 낭비되는 것같지는 않습니다!)

ChoJinmok commented on src/ControlMember/slice.js:
```diff
@@ -264,25 +255,65 @@ export function checkSignUpValid({ name, value }) {
-  function checkValid() {
-    if (name !== 'email' && name !== 'password') {
-      return false;
+export function checkInvalidMessageClear({ name, value }) {
```
전에 리뷰 드렸던 것처럼 함수의 역할과 함수의 이름이 일치하지 않는 거 같아요! 현재 함수의 역할은 회원가입 항목들 전체를 유효성 검사하는 것 같습니다!

ChoJinmok commented on src/ControlMember/ControlMemberContainer.jsx:
```diff
@@ -4,50 +4,57 @@ import { useDispatch, useSelector } from 'react-redux';
+  const handleCheckSignUpValid = useCallback(({
+    name, value,
+  }) => {
+    if (isButtonActive) return;
```
조건문을 `checkSignUpValid`액션에서 처리해주지 않고 함수에서 바로 처리해주신 이유가 있을까요?~ 정답이 없는 문제라 한번 고민해보시고 자신만의 기준을 찾아가시면 좋을 거 같아요!

xunxee replied:
isButtonActive가 true라면 dispatch를 실행하지 않고 바로 종료시키는 것이 좋다고 생각했는데,
이 부분에서 말씀해주신걸로 보면 액션에서 처리하는 것이 어떨까 하신 것 같네요.
이런 부분에서는 액션에서 조건문을 처리하면 어떤 점이 더 좋은가요??

ChoJinmok replied:
- 저같은 경우엔 컴포넌트에서는 단순히 상태를 받고, dispatch로 액션을 일으키게만 해주려고 하는 것 같아요! 컴포넌트라는 공간은 말그대로 데이터를 받아서 화면에 보여주는 로직정도만 있게하는거죠? 그리고 액션과 관련되고 상태를 관리하는 코드는 모두 slice로 이동시키는 것 같아요! 그리고 웬만한 로직은 무조건 slice에서 처리한다고 생각하면 기준을 잡기가 쉬워집니다. 워낙 다양한 상황이 많아서 어떨 때 그냥 컴포넌트에서 처리할지가 애매합니다.
- 제가 생각하는 handleCheckSignUpValid함수에 if문이 있으면 좋은 장점은 이미 isButtonActive라는 상태를 컴포넌트가 가지고 있고 활용할 수 있는 상황(slice에서는 getState로 상태를 다시 받아야한다.)이기 때문에 불필요하게 자원을 더 사용하지 않고 복잡한 로직이 아니기 때문에 '버튼이 활성화돼 있다면 함수를 종료한다'로 의미가 충분히 전달되는 것 같은 것 입니다!
- slice로 빼주는 장점은 가장 위에 말씀드린 것으로 설명이 될 거 같아요!
- 말씀하신 이유를 'isButtonActive가 true라면 액션이 실행되지 않는다'로 봤을때 이는 두가지 상황모두에 모두 적용할 수 있는 이유인 거 같아요! '액션이 아예 실행되지 않거나 실행하되 바로 종료시킨다'로 시점 정도가 달라지는 것 같습니다!
- 절대 '무조건 이렇게 하세요!'라고 말씀드리는게 아니구 코드를 작성했을 땐 목적이 있어야할 것 같아요! 코드를 제가 말씀드리면 무조건 고치건 가장 안좋고 이렇게 해도되고 저렇게 해도 되는 상황에서 이런 리뷰가 들어온다면 어떻게 코드를 작성하실건지 건땅 스스로 기준을 확고히 하시고 만약 큰 이유 없이 작성했다면 이 기회에 고민해보고 생각해보셨으면 하는 마음입니다.ㅎㅎ

ChoJinmok commented on src/List/ListContainer.jsx:
```diff
@@ -28,97 +30,117 @@ const Layout = styled.div({
-    if (product) {
-      onClickCategories(
-        `/search/products/${product}/${name}${search}`,
-      );
-
-      return;
+    url: {
+      category, product, sort, material,
+    },
+  } = listStates;
```
```javascript
const {
  url: {
    category, product, sort, material,
  },
} = useSelector(({ list }) => list);
```

이렇게 한번에 작성 가능할 것 같습니다!

xunxee replied:
한 번에 작성하면 아래에서 listStates를 사용하지 못하는데
한 번 더 useSelector를 아래서 사용하는 방법이 더 좋을까요?

ChoJinmok replied:
아래쪽에서 listStates가 사용되는군요?!

ChoJinmok commented on src/List/ListContainer.jsx:
```diff
@@ -28,97 +30,117 @@ const Layout = styled.div({
+      ]) => {
+        if (!categoryValue) return;
+
+        if (search.length) {
+          search.push(`&${categoryName}=${categoryValue}`);
+        }
+
+        if (!search.length) {
+          search.push(`?${categoryName}=${categoryValue}`);
+        }
```
이런식으로 하는 방식 참고하세요!

```javascript
if (search.length) {
  search.push(`&${categoryName}=${categoryValue}`);

  return;
}

search.push(`?${categoryName}=${categoryValue}`);
```

xunxee replied:
이런식으로 작성할 수 있겠군요!
감사합니다

ChoJinmok commented on src/List/ListContainer.jsx:
```diff
@@ -28,97 +30,117 @@ const Layout = styled.div({
+        if (categoryName === 'category') {
+          pathname.push(`/${categoryValue}`);
+        }
+      });
```
리뷰 드렸던 것처럼 지금과 같은 방식으로 했을 때 'product'보다 'category'가 먼저 반복문안으로 들어올 경우 순서가 바뀌는 에러가 발생할거에요!

ChoJinmok commented on src/List/ListContainer.jsx:
```diff
@@ -28,97 +30,117 @@ const Layout = styled.div({
+  } = listStates;
+
+  function checkUrl({ name, belong }) {
+    function returnEntries(kind) {
```
return보다는 make라고 표현하시는게 더 좋을 거 같아요!

ChoJinmok commented on src/List/ListContainer.jsx:
```diff
@@ -28,97 +30,117 @@ const Layout = styled.div({
-      belong: 'material',
-    }));
-  }, [dispatch]);
+    onClickCategories(checkUrl({ name, belong }));
```
코드의 줄 수가 줄어든다고 꼭 좋은 코드는 아닐 수 있을 거 같아요! 다른 사람이 읽어도 이해하기 쉽도록 작성하시면 좋을 거 같습니다!

```javascript
const url = checkUrl({ name, belong });

onClickCategories(url);
```

ChoJinmok commented on src/List/ListContainer.jsx:
```diff
@@ -28,97 +30,117 @@ const Layout = styled.div({
+    },
+  } = listStates;
+
+  function checkUrl({ name, belong }) {
```
checkUrl 함수도 역할이 url을 검사하는 것보다는 만드는 역할을 하는 거 같아요! makeUrl이 더 적절할 거 같아요!

ChoJinmok reviewed (commented):
control member보다 간단하게 membership은 어떠신가용?? 그리고 `HeaderContainer`에 `LoginPage` 컴포넌트 이름이 고쳐지지 않은 거 같습니당 ㅎㅎ

xunxee committed:
Refactor: ControlMember를 Membership으로 리팩터링

xunxee committed:
Refactor: Membership/slice.js의 checkInvalidMessageClear 함수의 이름을 checkMembershipValid로 리팩터링

xunxee committed:
Refactor: ListContainer 컴포넌트의 makeUrl 함수의 로직을 리팩터링
- makePathname 함수에서 정적으로 /product가 먼저 나오도록 로직을 리팩터링

xunxee committed:
Add: URL이 바뀌면 해당 URL에 맞게 changeUrlDataField 함수가 실행되는 액션을 구현
- urlPathname이 변하면 useEffect의 코드가 실행되어 dispatch를 발생시켜 store의 값을 최신화시킨다.

xunxee committed:
Add: List/slice.js의 changeUrlAllDataFields 함수의 액션을 구현 및 테스트
- url의 state 값을 한 번에 change하는 함수를 구현

xunxee committed:
Add: ListContainer 컴포넌트에서 queryString을 만드는 makeApiData 함수와 state를 저장하는 changeUrlData 함수를 구현
- 사용자가 주소를 직접 입력하여 새로고침이 일어나면 백엔드로 보낼 query string을 만드는 makeApiData 함수를 구현
- 입력한 주소의 데이터를 객체로 바꾸고 이를 state에 저장하는 changeUrlData 함수를 구현

xunxee committed:
Add: mockData를 받아오는 fetchMockData 함수의 액션을 테스트 및 구현

xunxee committed:
Add: GET 요청 이후 받아온 productList 데이터를 store에 저장하는 setProductList 함수의 액션을 테스트 및 구현

xunxee committed:
Add: List/slice.js의 loadProductList 함수의 액션을 테스트 및 구현

xunxee committed:
Add: ItemCard 컴포넌트의 기본적인 렌더링 사항을 테스트 및 구현

xunxee committed:
Add: List 페이지의 상품 목록을 한 줄에 3개씩 보이도록 스타일링을 추가

xunxee committed:
Add: ItemCard 컴포넌트에서 제품의 제목과 가격표의 기본적인 구성을 구현

xunxee committed:
Add: reset.js에 font 스타일을 추가

xunxee committed:
Add: ItemCard 컴포넌트를 hover시 color를 결정하는 함수의 로직을 구현 및 테스트
- product의 id를 가지고 color를 결정하는 함수 makeHoverColor를 추가

xunxee committed:
Add: style/Palette.js 스타일 파일을 추가
- 색상과 관련된 Palette.js를 모듈화

xunxee committed:
Add: ItemCard를 hover시 상품명과 이미지가 커지는 액션을 추가

xunxee committed:
Add: mockData/productList.json에 data를 추가
- img를 mainImage로 수정
- 상세 페이지에서 사용될 제품의 이미지 리스트를 추가

xunxee committed:
Add: ProductDetailPage의 기본적인 렌더링을 위한 액션을 테스트 및 구현

xunxee committed:
Add: List/slice.js의 setProduct 함수의 액션을 테스트 및 구현

xunxee committed:
Add: fetchMockProduct 함수의 액션을 테스트 및 구현

xunxee committed:
Add: List/slice.js의 loadProduct 함수의 액션을 테스트 및 구현

xunxee committed:
Add: ListContainer 컴포넌트의 makeQueryString 함수의 액션을 테스트 및 구현

xunxee committed:
Add: ListContainer 컴포넌트의 유효한 pathname이 아닐 경우의 예외처리 액션을 테스트 및 구현
- 기존 ListContainer 컴포넌트의 테스트 코드의 중복되는 코드를 리팩터링하기 위해 forEach를 사용하여 테스트 코드를 리팩터링

xunxee committed:
Add: /fixtures/MockData/product.js에 상품 상세 이미지들을 추가
- ProductDetailPage에서 사용될 MockData의 구성을 추가

xunxee committed:
Add: ProductDetailPage의 기본적인 렌더링을 위한 액션을 테스트

xunxee committed:
Add: styles/Palette.js에 color를 추가
- basicPurple, darkPurple, palePurple 추가

xunxee committed:
Add: ProductWrapper 컴포넌트에 화살표 아이콘을 추가
- SlideControlButton 컴포넌트를 추가

xunxee committed:
Add: ProductWrapper 컴포넌트의 화살표 버튼을 클릭시의 액션을 테스트 및 구현
- 슬라이드 구현시 state를 관리시 리덕스로 관리하는 것은 너무 무겁다는 생각이 들어 useState를 사용하여 상태를 관리

xunxee committed:
Add: ProductWrapper 컴포넌트의 handleClickLeftButton 함수의 액션을 테스트 및 구현

xunxee committed:
Add: UUID 모듈 사용을 위한 install

xunxee committed:
Add: productWrapper 컴포넌트에 UUID를 사용하여 key 값을 설정
- UUID를 사용한 key 값을 넣은 banners를 사용하기 위하여 코드를 리팩터링

xunxee committed:
Fix: public/mockData 파일의 line 에러를 수정

xunxee committed:
Add: fixtures/MockData/product.js에 banners를 추가
- Slide.test.jsx에서 banners를 사용하기 위해 해당 content를 추가

xunxee committed:
Add: ProductDetailContainer에서 초기 렌더링시 product의 imageList가 없을 경우의 액션을 테스트

xunxee committed:
Add: ProductDetailPage의 TODO 내용을 추가

xunxee committed:
Add: ProductWrapper 컴포넌트의 무한 슬라이드를 위한 액션을 테스트 및 구현

xunxee committed:
Add: Slide의 버튼 클릭시의 액션을 테스트 및 구현
